### PR TITLE
Rewrite custom templates article

### LINKED ISSUE #11876: Cross-platform option - clarification?
I see the comment about NuGetiser 3000.  However, the link points to an initial spec wiki page, with no instructions on use or how to obtain it.  The releases page in that Wiki doesn't contain it.  With a search, I find https://github.com/NuGet/NuGet.Build.Packaging/releases - but nothing since August 2017, and it's a source code download.

Is there any more guidance you could share about how to actually do a "nuget pack" cross-platform?  I "had" just removed my dependency on mono on my build system... do I really need to install it again and use nuget.exe?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4e7deb06-37bb-b659-7cf3-9c47a91e065e
* Version Independent ID: d6e6f989-3956-6a0e-5e9c-fb3d1f9bd751
* Content: [Custom templates for dotnet new - .NET Core CLI](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates#packing-a-template-into-a-nuget-package-nupkg-file)
* Content Source: [docs/core/tools/custom-templates.md](https://github.com/dotnet/docs/blob/master/docs/core/tools/custom-templates.md)
* Product: **dotnet-core**
* Technology: **dotnet-cli**
* GitHub Login: @guardrex
* Microsoft Alias: **mairaw**

### LINKED ISSUE #12002: dotnet new -i FILE_SYSTEM_PATH / dotnet new -u undocumented behaviour
The page here says, "The FILE_SYSTEM_DIRECTORY is the project folder containing the project and the .template.config folder".

However, it looks like I can `dotnet new -i ANY_FOLDER_HERE`, and all subfolders will be searched, and any templates found will be installed?

Furthermore, I cannot `dotnet new -i X` and then `dotnet new -u X`.  For uninstallation, the path must be the full/absolute/canonical path, not a relative one.

Further-furthermore, templates automatically installed from a parent folder can only be uninstalled using the canonical path of the parent folder - not their individual folders.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4e7deb06-37bb-b659-7cf3-9c47a91e065e
* Version Independent ID: d6e6f989-3956-6a0e-5e9c-fb3d1f9bd751
* Content: [Custom templates for dotnet new - .NET Core CLI](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates)
* Content Source: [docs/core/tools/custom-templates.md](https://github.com/dotnet/docs/blob/master/docs/core/tools/custom-templates.md)
* Product: **dotnet-core**
* Technology: **dotnet-cli**
* GitHub Login: @guardrex
* Microsoft Alias: **mairaw**

### PR DESCRIPTION
**Pull request by Thraka**
Fixes #11876
Fixes #12002
Contributes to #12606

Described how to create a csproj that is supported with `dotnet pack` to create a template. Tested on Windows and Linux. Update to the tutorial will follow.

### CODE REVIEW

File: docs/core/tools/custom-templates.md [hunk 1]
```diff
@@ -4,33 +4,44 @@ description: Learn about custom templates for any type of .NET project or files.
 author: mairaw
 ms.date: 08/11/2017
 ---
+
 # Custom templates for dotnet new
 
-The [.NET Core SDK](https://www.microsoft.com/net/download/core) comes with many templates pre-installed to use with the [`dotnet new` command](dotnet-new.md). Starting with .NET Core 2.0, you can create your own custom templates for any type of project, such as an app, service, tool, or class library. You can even create a template that outputs one or more independent files, such as a configuration file.
+The [.NET Core SDK](https://www.microsoft.com/net/download/core) comes with many templates already installed and ready for you to use. The [`dotnet new` command](dotnet-new.md) isn'tonly the way to use a template, but also how to install and uninstall templates. Starting with .NET Core 2.0, you can create your own custom templates for any type of project, such as an app, service, tool, or class library. You can even create a template that outputs one or more independent files, such as a configuration file.
```
Comment (BillWagner) [inline docs/core/tools/custom-templates.md]:
```suggestion
The [.NET Core SDK](https://www.microsoft.com/net/download/core) comes with many templates already installed and ready for you to use. The [`dotnet new` command](dotnet-new.md) isn't only the way to use a template, but also how to install and uninstall templates. Starting with .NET Core 2.0, you can create your own custom templates for any type of project, such as an app, service, tool, or class library. You can even create a template that outputs one or more independent files, such as a configuration file.
```

File: docs/core/tools/custom-templates.md [hunk 2]
```diff
@@ -4,33 +4,44 @@ description: Learn about custom templates for any type of .NET project or files.
...
 
 - The template engine doesn't require you to inject special tokens into your project's source code.
 - The code files aren't special files or modified in any way to work with the template engine. So, the tools you normally use when working with projects also work with template content.
 - You build, run, and debug your template projects just like you do for any of your other projects.
-- You can quickly create a template from an existing project just by adding a *template.json* configuration file to the project.
+- You can quickly create a template from an existing project just by adding a *./.template.config/template.json* configuration file to the project.
+
+Files and folders stored in the template aren't limited to formal .NET project types. Source files and folders may consist of any content that you wish to create when the template is used, even if the template engine produces just one file for its output.
```
Comment (BillWagner) [inline docs/core/tools/custom-templates.md]: nit:

```suggestion
Files and folders stored in the template aren't limited to formal .NET project types. Source files and folders may consist of any content that you wish to create when the template is used, even if the template engine produces just one file as its output.
```

File: docs/core/tools/custom-templates.md [hunk 3]
```diff
@@ -40,105 +51,193 @@ The *template.json* file is placed in a *.template.config* folder in the root di
...
+| `classifications` | array(string) | Zero or more characteristics of the template that a user might use to find the template when searching for it. The classifications also appear in the *Tags* column when it appears in a list of templates produced by using the `dotnet new -l|--list` command. |
 | `identity`        | string        | A unique name for this template. |
 | `name`            | string        | The name for the template that users should see. |
-| `shortName`       | string        | A default shorthand for selecting the template that applies to environments where the template name is specified by the user, not selected via a GUI. For example, the short name is useful when using templates from a command prompt with CLI commands. |
+| `shortName`       | string        | A default shorthand name for selecting the template that applies to environments where the template name is specified by the user, not selected via a GUI. For example, the short name is useful when using templates from a command prompt with CLI commands. |
```
Comment (BillWagner) [inline docs/core/tools/custom-templates.md]: Do we have naming conventions we recommend for these? If so, we should add them.

File: docs/core/tools/custom-templates.md [hunk 4]
```diff
@@ -40,105 +51,193 @@ The *template.json* file is placed in a *.template.config* folder in the root di
...
+Provide the path to a *.nupkg* NuGet package file.
+
 ```console
 dotnet new -i <PATH_TO_NUPKG_FILE>
 ```
 
 ### To install a template from a file system directory
 
-The `FILE_SYSTEM_DIRECTORY` is the project folder containing the project and the *.template.config* folder:
+Templates can be installed by from a template folder, such as the *mytemplate1* folder from the example above. Specify the folder path of the *.template.config* folder. The path to the template directory does not need to be absolute. However, an absolute path is required to uninstall a template that is installed from a folder.
```
Comment (BillWagner) [inline docs/core/tools/custom-templates.md]:
```suggestion
Templates can be installed from a template folder, such as the *mytemplate1* folder from the example above. Specify the folder path of the *.template.config* folder. The path to the template directory does not need to be absolute. However, an absolute path is required to uninstall a template that is installed from a folder.
```

**Review by BillWagner — APPROVED** [review]
This looks great @Thraka 

I had a few small comments for you to consider before you :shipit:

File: docs/core/tools/custom-templates.md (hunk 3 above)
Comment (Thraka) [reply docs/core/tools/custom-templates.md]: Not seeing any. The schema just accepts string/array. So apparently multiples are allowed. I'll play with that 😁

**Comment by Thraka** [comment]
@BillWagner question on the style of presenting files. Generally I only see `*csproj*` for a file, but I thought it would look better if it was `*.csproj*` what do you think? It freaks out acrolinx so I was going to file a request that excluded the pattern.

**Pull request merged**